File: src/lowlevel/server/icefsoperators/icefsReadLink.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-18 11:54:31
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsReadLink.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"

	"golang.org/x/sys/unix"
)

func (s *IcefsServer) doIcefsReadLink(fakeInode uint64) (status int32, path []byte, readN int) {
	var err error
	s.inodeCacheLock.RLock()
	inode := s.getIcefsInode(fakeInode)
	if inode == nil {
		s.inodeCacheLock.RUnlock()
		status = icefserror.ICEFS_BUG_ERR
		return
	}

	path = make([]byte, unix.PathMax+1)

	inode.inodeLock.RLock()
	s.inodeCacheLock.RUnlock()
	readN, err = unix.Readlinkat(inode.fd, "", path)
	inode.inodeLock.RUnlock()
	status = icefserror.IcefsStdErrno(err)
	return
}

func (s *IcefsGRpcServer) DoIcefsReadLink(ctx context.Context, req *pb.IcefsReadLinkReq) (*pb.IcefsReadLinkRes, error) {
	var res pb.IcefsReadLinkRes
	var path []byte
	var readN int

	res.Status, path, readN = s.server.doIcefsReadLink(req.Inode)
	if (res.Status == icefserror.ICEFS_EOK) && readN == len(path) {
		res.Path = string(path)
	}

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsReadLink(ctx context.Context, req *icefsthrift.IcefsReadLinkReq) (*icefsthrift.IcefsReadLinkRes, error) {
	var res icefsthrift.IcefsReadLinkRes
	var path []byte
	var readN int

	res.Status, path, readN = s.server.doIcefsReadLink(uint64(req.Inode))
	if (res.Status == icefserror.ICEFS_EOK) && readN == len(path) {
		res.Path = string(path)
	}

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsReadLink.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:58
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsReadLink.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsReadLink(fuse_req_t fuseReq, fuse_ino_t inode) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsReadLinkReq req;
      icefsgrpc::IcefsReadLinkRes res;
      grpc::ClientContext ctx;

      req.set_inode(inode);

      grpc::Status status = gRpcClient->DoIcefsReadLink(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_readlink(fuseReq, res.path().c_str());
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsReadLinkReq req;
      icefsthrift::IcefsReadLinkRes res;

      req.__set_inode(inode);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsReadLink(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      if (!res.status) {
        fuse_reply_readlink(fuseReq, res.path.c_str());
      } else {
        fuse_reply_err(fuseReq, res.status);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}